desmod/simulation.py

```diff
@@ -183,6 +183,7 @@ def simulate(config, top_type, env_type=SimEnvironment, reraise=True,
                 except BaseException as e:
                     env.tracemgr.trace_exception()
                     result['sim.exception'] = repr(e)
+                    top.post_fail()
```

Comment:
If either `top_type.pre_init()` or `top_type()` fail, then the `top` local will not exist when entering this exception handler. This would lead to a `NameError` here where we reference `top`.

To retain this strategy of calling `top.post_fail()`, yet another nested `try`/`except` would have to be added after `top` is initialized.

The new indentation level would be awkward, but it also means that the scope of error catching would exclude `top_type.__init__()` and `top_type.pre_init()`. Anyone writing a `Component` that implements a `post_fail_hook()` would need to comprehend this limitation. It also means that there would not be a solution to handling exceptions raised during `top.pre_init()` or `top.__init__()`.

A further complication is that `post_fail_hook()` methods might be called either during or after `post_simulate()`. Thus `post_fail_hook()` implementations would have to take care to idempotently tear-down any resources that might have been affected by `post_simulate()`.